src/quick-lint-js/parse.h

```diff
@@ -2460,6 +2460,71 @@ class parser {
     v.visit_exit_with_scope();
   }
 
+  bool is_side_effect_free(expression *ast) {
```

Comment:
Nit: Maybe it'd be better to invert the meaning of this function and call it `has_potential_side_effects`? `!is_side_effect_free(ast)` sounds like a double negative to me.